spinn_machine/json_machine.py

```diff
@@ -0,0 +1,338 @@
+
+    def __str__(self):
+        return "[JsonMachine: max_x={}, max_y={}, n_chips={}]".format(
+            self._max_chip_x, self._max_chip_y, self.n_chips)
+
+    @staticmethod
+    def _int_value(value):
+        if value < JAVA_MAX_INT:
+            return value
+        else:
+            return JAVA_MAX_INT
+
+    @staticmethod
+    def _find_virtual_links(machine):
+        """ Find all the virtual links and their inverse.
+
+        As these may well go to an unexpected source
+
+        :param machine: Machine to convert
+        :return: Map of Chip to list of virtual links
+        """
+        virtual_links_dict = defaultdict(list)
+        for chip in machine._virtual_chips:
+            # assume all links need special treatment
+            for link in chip.router.links:
+                virtual_links_dict[chip].append(link)
+                # Find and save inverse link as well
+                inverse_id = (
+                    (link.source_link_id + Router.MAX_LINKS_PER_ROUTER // 2)
+                    % Router.MAX_LINKS_PER_ROUTER)
+                destination = machine.get_chip_at(
+                    link.destination_x, link.destination_y)
+                inverse_link = destination.router.get_link(inverse_id)
+                assert(inverse_link.destination_x == chip.x)
+                assert(inverse_link.destination_y == chip.y)
+                virtual_links_dict[destination].append(inverse_link)
+        return virtual_links_dict
+
+    @staticmethod
+    def _describe_chip(chip, std, eth, virtual_links_dict):
+        """ Produce a JSON-suitable description of a single chip.
+
+        :param chip: The chip to describe.
+        :param std: The standard chip resources.
+        :param eth: The standard ethernet chip resources.
+        :param virtual_links_dict: Where the virtual links are.
+        :return: Description of chip that is trivial to serialize as JSON.
+        """
+        details = OrderedDict()
+        details["cores"] = chip.n_processors
+        if chip.nearest_ethernet_x is not None:
+            details["ethernet"] =\
+                [chip.nearest_ethernet_x, chip.nearest_ethernet_y]
+
+        dead_links = []
+        for link_id in range(0, Router.MAX_LINKS_PER_ROUTER):
+            if not chip.router.is_link(link_id):
+                dead_links.append(link_id)
+        if dead_links:
+            details["deadLinks"] = dead_links
+
+        if chip in virtual_links_dict:
+            links = []
+            for link in virtual_links_dict[chip]:
+                link_details = OrderedDict()
+                link_details["sourceLinkId"] = link.source_link_id
+                link_details["destinationX"] = link.destination_x
+                link_details["destinationY"] = link.destination_y
+                links.append(link_details)
+            details["links"] = links
+
+        exceptions = OrderedDict()
+        router_entries = JsonMachine._int_value(
```

Comment:
Ive read this code up above, thus cloned code. should be able to make that a method for both calls

Reply:
yea but no